airflow/contrib/task_runner/cgroup_task_runner.py

```diff
@@ -0,0 +1,178 @@
+        node = trees.Tree().root
+        path_split = path.split("/")
+        for path_element in path_split:
+            name_to_node = {x.name: x for x in node.children}
+            if path_element not in name_to_node:
+                self.logger.warn("Cgroup does not exist: {}"
+                                 .format(path))
+                return
+            else:
+                node = name_to_node[path_element]
+        # node is now the leaf node
+        parent = node.parent
+        self.logger.debug("Deleting cgroup {}/{}".format(parent, node.name))
+        parent.delete_cgroup(node.name)
+
+    def start(self):
+        # Use bash if it's already in a cgroup
+        cgroups = self._get_cgroup_names()
+        if cgroups["cpu"] != "/" or cgroups["memory"] != "/":
+            self.logger.debug("Already running in a cgroup (cpu: {} memory: {} so "
+                              "not creating another one"
+                              .format(cgroups.get("cpu"),
+                                      cgroups.get("memory")))
+            self.process = self.run_command(['bash', '-c'], join_args=True)
+            return
+
+        # Create a unique cgroup name
+        cgroup_name = "airflow/{}/{}".format(datetime.datetime.now().
+                                             strftime("%Y-%m-%d"),
+                                             str(uuid.uuid1()))
+
+        self.mem_cgroup_name = "memory/{}".format(cgroup_name)
+        self.cpu_cgroup_name = "cpu/{}".format(cgroup_name)
+
+        # Get the resource requirements from the task
+        task = self._task_instance.task
+        resources = task.resources
+        cpus = resources.cpus.qty
+        self._cpu_shares = cpus * 1024
+        self._mem_mb_limit = resources.ram.qty
+
+        # Create the memory cgroup
+        mem_cgroup_node = self._create_cgroup(self.mem_cgroup_name)
+        self._created_mem_cgroup = True
+        if self._mem_mb_limit > 0:
+            self.logger.debug("Setting {} with {} MB of memory"
+                              .format(self.mem_cgroup_name, self._mem_mb_limit))
+            mem_cgroup_node.controller.limit_in_bytes = self._mem_mb_limit * 1024 * 1024
+
+        # Create the CPU cgroup
+        cpu_cgroup_node = self._create_cgroup(self.cpu_cgroup_name)
+        self._created_cpu_cgroup = True
+        if self._cpu_shares > 0:
+            self.logger.debug("Setting {} with {} CPU shares"
+                              .format(self.cpu_cgroup_name, self._cpu_shares))
+            cpu_cgroup_node.controller.shares = self._cpu_shares
+
+        # Start the process w/ cgroups
+        self.logger.debug("Starting task process with cgroups cpu,memory:{}"
+                          .format(cgroup_name))
+        self.process = self.run_command(
+            ['cgexec', '-g', 'cpu,memory:{}'.format(cgroup_name)]
+        )
+
+    def return_code(self):
+        return_code = self.process.poll()
+        if return_code == 137:
```

Comment:
So there was an improved method for detecting when an OOM happens by monitoring the control file in the cgroup fs. I had it working at one point with 

https://github.com/plypaul/airflow/blob/e144e4d41996300ffa93947f136eab7785b114ed/airflow/contrib/task_runner/cgroup_task_runner.py#L43 

but later there were some issues installing the python `butter` package and `libseccomp-dev` on some hosts for some reason. I wasn't able to track down the root cause of the package install failures, but we might want to revisit that approach at some other point.